### PR TITLE
avi_obj_vs AviVsVipBuild cleanup, fix AviVsVipBuild static IP usecases in public clouds

### PR DESCRIPTION
**Pull request by shubhamavi**
This includes/extends static IP vsvip configuration to public cloud types. Prior to this public cloud and staticIP vsvip implementation was exclusive of each other.
This also involves a major cleanup in the AviVsVipBuild function, to make it more readable, trimmed it down a bit for easier logical workflow.
I have intentionally kept the two commits separated out, where the first one just focuses on providing better readability, and the second one fixes the bug reported AV-102214

Since there have been multiple authors for this block of code, lets merge it when all you guys approve it @parimanur @monotosh-avi @sudswasavi 
Please try checking for all the use cases, including the fix introduced in the second commit

### CODE REVIEW

**Comment by shubhamavi** [comment]
build ako

File: internal/rest/avi_obj_vs.go [hunk 1]
```diff
@@ -632,88 +642,83 @@ func (rest *RestOperations) AviVsVipBuild(vsvip_meta *nodes.AviVSVIPNode, cache_
...
+		}
 	} else {
-		auto_alloc := true
-		var vips []*avimodels.Vip
-
-		// all vsvip models would have auto_alloc set to true even in case of static IP programming
-		vip := avimodels.Vip{AutoAllocateIP: &auto_alloc}
-		networkRef := lib.GetNetworkName()
-		if lib.IsPublicCloud() && lib.GetNetworkName() != "" {
+		// all vsvip models would have AutoAllocateIP set to true even in case of static IP programming
```
Comment (monotosh-avi) [inline internal/rest/avi_obj_vs.go:670]: Is this comment required again ?

File: internal/rest/avi_obj_vs.go [hunk 2]
```diff
@@ -632,88 +642,83 @@ func (rest *RestOperations) AviVsVipBuild(vsvip_meta *nodes.AviVSVIPNode, cache_
...
+				},
+			}
+		}
+
+		// configuring static IP, from gateway.Addresses (advl4) and service.loadBalancerIP (l4)
+		if vsvip_meta.IPAddress != "" {
+			vip.IPAddress = &avimodels.IPAddr{Type: &ipType, Addr: &vsvip_meta.IPAddress}
+		}
+
+		// selecting network with user input, in case user inputt is not provided AKO relies on
```
Comment (monotosh-avi) [inline internal/rest/avi_obj_vs.go:703]: inputt -> input

**Comment by shubhamavi** [comment]
build ako

**Review by monotosh-avi — COMMENTED** [review]
looks good to me.

File: internal/rest/avi_obj_vs.go [hunk 3]
```diff
@@ -632,88 +642,83 @@ func (rest *RestOperations) AviVsVipBuild(vsvip_meta *nodes.AviVSVIPNode, cache_
...
+		// all vsvip models would have AutoAllocateIP set to true even in case of static IP programming
+		vip := avimodels.Vip{
+			VipID:          &vipId,
+			AutoAllocateIP: &autoAllocate,
+		}
 
+		// setting IPAMNetworkSubnet.Subnet value in case subnetCIDR is provided
+		if lib.GetSubnetPrefix() == "" || subnetAddress == "" {
+			utils.AviLog.Warnf("Incomplete values provided for subnetIP, will not use IPAMNetworkSubnet in vsvip")
+		} else if lib.IsPublicCloud() {
```
Comment (parimanur) [inline internal/rest/avi_obj_vs.go:679]: this can further be simplified with this condition `lib.GetCloudType() == lib.CLOUD_GCP`

**Review by parimanur — COMMENTED** [review]
LGTM with one minor change.

File: internal/rest/avi_obj_vs.go (hunk 1 above)
Comment (shubhamavi) [reply internal/rest/avi_obj_vs.go]: done

File: internal/rest/avi_obj_vs.go (hunk 2 above)
Comment (shubhamavi) [reply internal/rest/avi_obj_vs.go]: done

File: internal/rest/avi_obj_vs.go (hunk 3 above)
Comment (shubhamavi) [reply internal/rest/avi_obj_vs.go]: done

**Comment by shubhamavi** [comment]
build ako

**Comment by shubhamavi** [comment]
build ako

**Comment by shubhamavi** [comment]
build ako

**Review by sudswasavi — APPROVED** [review]

**Review by parimanur — APPROVED** [review]

**Comment by gurunath-avi** [comment]
build ako

**Comment by shubhamavi** [comment]
build ako

**Pull request merged**